### PR TITLE
Re-sync in case of stale offset

### PR DESCRIPTION
**Pull request by rodireich**
## What
With this change mysql will now correctly do a re-sync in case a saved CDC offset is found to be stale, and mysql is configured to go to resync rather than fail.
image

## How
CDCPartitionsCreator goes into the following flow:
1. Round 1: All streams are rolled back to an initial state so the next sync attempt will start over.
2. Round 2: A transient exception is thrown so platform will kick off another sync attempt.
3. On the next attempt all streams are rebuilt from scratch.

## Review guide
1. CdcPartitionsCreator - new handling of re-sync flow in 2 rounds.
2. StateManager - adding a reset streams state to be able to roll back and start over.
3. MySqlDebeziumOperations - Identify the stale offset + configured to re-sync and trigger the new flow.

### CODE REVIEW

File: airbyte-integrations/connectors/source-mysql/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcPartitionFactory.kt [hunk 1]
```diff
@@ -158,7 +157,10 @@ class MysqlJdbcPartitionFactory(
         val stream: Stream = streamFeedBootstrap.feed
         val streamState: DefaultJdbcStreamState = streamState(streamFeedBootstrap)
         val opaqueStateValue: OpaqueStateValue =
-            streamFeedBootstrap.currentState ?: return coldStart(streamState)
+            when (streamFeedBootstrap.currentState?.isEmpty) {
+                false -> streamFeedBootstrap.currentState!!
+                else -> return coldStart(streamState)
+            }
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-mysql/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcPartitionFactory.kt:163]: This fixes a bug:
When sync that includes a number of feeds with some that haven't started yet, during checkpoint those feeds will have a stream state of `{}`.
Stopping and restarting the sync need to treat them as a cold start. before this fix those `{}` feeds didn't do a snapshot.

**Comment by rodireich** [comment]
While verifying this pr I noticed another bug that is not directly related to resync but has to do with how empty tables save their state. 

I'm fixing it but please take a look meantime

File: airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io/airbyte/cdk/read/cdc/CdcPartitionsCreator.kt [hunk 2]
```diff
@@ -39,6 +42,12 @@ class CdcPartitionsCreator<T : Comparable<T>>(
     }
 
     override suspend fun run(): List<PartitionReader> {
+        if (CDCNeedsRestart) {
+            globalLockResource.markCdcAsComplete()
+            throw TransientErrorException(
+                "Saved offset no longer present on the server, Airbyte is going to trigger a sync from scratch."
```
Comment (theyueli) [inline airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io/airbyte/cdk/read/cdc/CdcPartitionsCreator.kt:48]: nit: "presents"

**Review by theyueli — APPROVED** [review]

**Comment by postamar** [comment]
This PR breaks the CAT tests for source-oracle. This PR is also missing tests. We shouldn't revert it but this definitely needs to be reworked properly.

**Pull request merged**